astropy/utils/tests/test_data.py

```diff
@@ -553,10 +1422,101 @@ def read(self, length=None):
     real_length = 1023
     fn = download_file(TESTURL, cache=False)
     with open(fn, "rb") as f:
-        assert f.read() == b"a"*real_length
+        assert f.read() == b"a" * real_length
 
     report_length = None
     real_length = 1023
     fn = download_file(TESTURL, cache=False)
     with open(fn, "rb") as f:
-        assert f.read() == b"a"*real_length
+        assert f.read() == b"a" * real_length
+
+
+def test_can_make_directories_readonly(tmpdir):
+    try:
+        with readonly_dir(tmpdir):
+            assert is_dir_readonly(tmpdir)
+    except AssertionError:
+        if hasattr(os, "geteuid") and os.geteuid() == 0:
+            pytest.skip(
+                "We are root, we can't make a directory un-writable with chmod."
+            )
+        elif platform.system() == "Windows":
+            pytest.skip(
+                "It seems we can't make a driectory un-writable under Windows "
+                "with chmod, in spite of the documentation."
+            )
+        else:
+            raise
+
+
+def test_can_make_files_readonly(tmpdir):
+    fn = tmpdir / "test"
+    c = "contents\n"
+    with open(fn, "w") as f:
+        f.write(c)
+    with readonly_dir(tmpdir):
+        try:
+            with open(fn, "w+") as f:
+                f.write("more contents\n")
+        except PermissionError:
+            pass
+        else:
+            if hasattr(os, "geteuid") and os.geteuid() == 0:
+                pytest.skip("We are root, we can't make a file un-writable with chmod.")
+    assert get_file_contents(fn) == c
+
+
+def test_read_cache_readonly(readonly_cache):
+    assert cache_contents() == readonly_cache
+
+
+def test_download_file_cache_readonly(readonly_cache):
+    for u in readonly_cache:
+        f = download_file(u, cache=True)
+        assert f == readonly_cache[u]
+
+
+def test_download_file_cache_readonly_cache_miss(readonly_cache, valid_urls):
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    assert get_file_contents(f) == c
+    assert not is_url_in_cache(u)
+
+
+def test_download_file_cache_readonly_update(readonly_cache):
+    for u in readonly_cache:
+        f = download_file(u, cache="update")
+        assert f != readonly_cache[u]
+        assert compute_hash(f) == os.path.basename(readonly_cache[u])
+
+
+def test_check_download_cache_works_if_readonly(readonly_cache):
+    check_download_cache(check_hashes=True)
+
+
+def test_read_cache_fake_readonly(fake_readonly_cache):
```

Comment:
Why the need for "real" and "fake" readonly caches in the tests?

Reply:
I can't make directories readonly on Windows, and I can't make anything readonly on CircleCI, so I can't necessarily use real tests. But I don't totally trust monkeypatch to simulate readonliness either so I don't want to rely on those ("fake") tests either.

Clearly I should add some explanation to the test suite.